contracts/origination/OriginationControllerSTIRFRY.sol

```diff
@@ -0,0 +1,268 @@
+     * @param _vaultFactory                 The address of the vault factory.
+     */
+    constructor(
+        address _originationHelpers,
+        address _loanCore,
+        address _feeController,
+        address _vaultFactory
+    ) OriginationControllerBase(_originationHelpers, _loanCore) {
+        if (_feeController == address(0)) revert OCS_ZeroAddress("feeController");
+        if (_vaultFactory == address(0)) revert OCS_ZeroAddress("vaultFactory");
+
+        feeController = IFeeController(_feeController);
+        vaultFactory = IVaultFactory(_vaultFactory);
+    }
+
+    // ======================================= LOAN ORIGINATION =========================================
+
+    /**
+     * @notice Initializes a new STIRFRY loan with Loan Core.
+     *
+     * @notice If item predicates are passed, they are used to verify collateral.
+     *
+     * @dev The caller must be a borrower or lender, or approved by a borrower or lender.
+     * @dev The external signer must be a borrower or lender, or approved by a borrower or lender.
+     * @dev The external signer must come from the opposite side of the loan as the caller.
+     *
+     * @param loanTerms                     The terms agreed by the lender and borrower.
+     * @param stirfryData                   The stirfry data to initialize the loan with.
+     * @param borrower                      Address of the borrower.
+     * @param lender                        Address of the lender.
+     * @param sig                           The loan terms signature, with v, r, s fields, and possible extra data.
+     * @param sigProperties                 Signature nonce and max uses for this nonce.
+     * @param itemPredicates                The predicate rules for the items in the bundle.
+     *
+     * @return loanId                       The unique ID of the new loan.
+     */
+    function initializeStirfryLoan(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        StirfryData calldata stirfryData,
+        address borrower,
+        address lender,
+        Signature calldata sig,
+        SigProperties calldata sigProperties,
+        LoanLibrary.Predicate[] calldata itemPredicates
+    ) external returns (uint256 loanId) {
+        // input validation
+        originationHelpers.validateLoanTerms(loanTerms);
+
+        address vaultAddress = vaultFactory.instanceAt(loanTerms.collateralId);
+        _validateStirfryTerms(loanTerms, stirfryData, vaultAddress);
+
+        {
+            // signature validation
+            Side neededSide = isSelfOrApproved(borrower, msg.sender) ? Side.LEND : Side.BORROW;
+
+            address signingCounterparty = neededSide == Side.LEND ? lender : borrower;
+            address callingCounterparty = neededSide == Side.LEND ? borrower : lender;
+
+            (bytes32 sighash, address externalSigner) = _recoverSignature(loanTerms, sig, sigProperties, neededSide, signingCounterparty, itemPredicates);
```

Comment:
Wouldn't `vaultedCurrency` need to be part of the signature? I think you need a new typehash for this; otherwise, let's say `stirfryPairs` allowed both sUSDe <> USDC _and_"uni v2 USDC<>USDT LP Token" <> USDC.

If I wanted to deliver fixed interest in USDC, then `terms.payableCurrency` would be USDC. Couldn't the counterparty swapping for fixed (the "lender" in terms of protocol designation) then call this function, but specify the _other_ pair, the one against LP Tokens? It's not in my signature, so it seems like they could. Then all the sudden I'd be speculating on Uni V2 LP interest rates instead of sUSDe.

Reply:
Made this change, good call. Rearranged the OriginationHelpers library contract a bit to add in the type hash (and encoding fn) and clean up a few things (moved `isApprovedForContract()` to OCBase). Signature recovery I put in the OCInterestRateSwap contract and kept it out of the OCBase contract.